File: tensorrt_llm/_torch/pyexecutor/generate_request_options.py
from dataclasses import dataclass
from typing import List, Optional, Tuple

import torch

from tensorrt_llm.bindings import ModelConfig, WorldConfig, SamplingConfig
from tensorrt_llm.bindings.executor import DecodingConfig
from tensorrt_llm.bindings.internal.batch_manager import DecoderInputBuffers, LlmRequest
from tensorrt_llm.bindings.internal.runtime import Request


@dataclass
class GenerateRequestOptions:
    """
    Python implementation of GenerateRequestOptions.
    Implements the logic in `generateRequestOptions.cpp`.
    """
    speculative_decoding_fast_logits: bool
    is_leader_in_orch_mode: bool
    is_normalize_log_probs: bool

    def __call__(
        self,
        model_config: ModelConfig,
        world_config: WorldConfig,
        decoding_config: DecodingConfig,
        context_requests: List[LlmRequest],
        logits_type: torch.dtype,
        input_buffers: DecoderInputBuffers,
    ) -> Tuple[torch.Tensor, List[Request], List[SamplingConfig]]:
        """Process context requests and prepare decoder requests.

        Args:
            model_config: Model configuration
            world_config: World configuration
            decoding_config: Decoding configuration
            context_requests: List of context requests to process
            logits_type: Data type for logits
            input_buffers: Input buffers for decoder

        Returns:
            Tuple containing:
            - Batch slots view tensor
            - List of decoder requests
            - List of sampling configs
        """
        batch_size = 0
        decoder_input_size = 0

        # Calculate batch size and decoder input size
        for llm_req in context_requests:
            req_tokens = llm_req.get_tokens(0)
            if llm_req.is_last_context_chunk():
                decoder_input_size += len(req_tokens)
                batch_size += 1

        # Resize input buffer
        input_buffers.inputs_ids = torch.empty(decoder_input_size, dtype=torch.int32, device='cuda')

        # Create batch slots view
        batch_slots_view = input_buffers.setup_batch_slots[:batch_size]

        decoder_requests = []
        sampling_configs = []

        batch_idx = 0
        input_offset = 0

        # Process each context request
        for llm_req in context_requests:
            if not llm_req.is_last_context_chunk():
                continue

            prompt_len = llm_req.prompt_len
            req_tokens = torch.tensor(llm_req.get_tokens(0))
            assert len(req_tokens) == prompt_len

            # Create input view and copy tokens
            input_view = input_buffers.inputs_ids[input_offset:input_offset + prompt_len]
            input_view.copy_(req_tokens, non_blocking=True)

            # Create decoder request
            decoder_request = Request(
                ids=input_view,
                input_len=prompt_len,
                max_new_tokens=llm_req.max_new_tokens,
                end_id=llm_req.end_id
            )

            # Set sampling config
            llm_req.sampling_config.normalize_log_probs = self.is_normalize_log_probs

            # Handle speculative decoding
            if model_config.speculative_decoding_mode.is_draft_tokens_external:
                if llm_req.has_draft_tokens():
                    draft_tokens = llm_req.draft_tokens
                    decoder_request.draft_tokens = draft_tokens.clone(memory_format=torch.contiguous_format).pin_memory()
                    
                    draft_logits = llm_req.draft_logits
                    if draft_logits is not None:
                        decoder_request.draft_logits = self._retrieve_draft_logits(
                            model_config, world_config, draft_logits)
                    
                    decoder_request.generated_tokens_per_engine_step = len(draft_tokens) + 1
                else:
                    decoder_request.generated_tokens_per_engine_step = 1
            elif not model_config.speculative_decoding_mode.is_none:
                decoder_request.generated_tokens_per_engine_step = model_config.get_max_decoding_tokens()

            # Handle Medusa speculative decoding
            if model_config.speculative_decoding_mode.is_medusa:
                # NOTE: The logic of Medusa requires a RuntimeBuffers object.
                raise NotImplementedError("Medusa speculative decoding is not currently supported in TRTLLMDecoder.")

            # Handle lookahead decoding
            elif model_config.speculative_decoding_mode.is_lookahead_decoding:
                decoder_request.lookahead_runtime_config = (
                    llm_req.get_lookahead_config() or decoding_config.get_lookahead_decoding_config()
                )

            # Handle explicit draft tokens
            elif model_config.speculative_decoding_mode.is_explicit_draft_tokens:
                decoder_request.dtype = model_config.get_data_type()

            # Handle Eagle speculative decoding
            elif model_config.speculative_decoding_mode.is_eagle:
                decoder_request.eagle_config = (
                    llm_req.get_eagle_config() or decoding_config.get_eagle_config()
                )

            # Handle embedding bias
            if llm_req.embedding_bias is not None:
                decoder_request.embedding_bias = self._get_embedding_bias(
                    logits_type, llm_req.embedding_bias)

            # Handle bad words list
            if llm_req.bad_words_list is not None:
                decoder_request.bad_words_list = llm_req.bad_words_list.clone().squeeze_(0)

            # Handle stop words list
            if llm_req.stop_words_list is not None:
                decoder_request.stop_words_list = llm_req.stop_words_list.clone().squeeze_(0)

            # Set batch slot and add to lists
            batch_slots_view[batch_idx] = llm_req.seq_slot
            decoder_requests.append(decoder_request)
            sampling_configs.append(llm_req.sampling_config)

            input_offset += prompt_len
            batch_idx += 1

        return batch_slots_view, decoder_requests, sampling_configs

    def _retrieve_draft_logits(
        self,
        model_config: ModelConfig,
        world_config: WorldConfig,
        tensor: torch.Tensor,
    ) -> torch.Tensor:
        """Retrieve draft logits for speculative decoding.

        Args:
            model_config: Model configuration
            world_config: World configuration
            tensor: Input tensor containing draft logits

        Returns:
            Processed draft logits tensor
        """
        raise NotImplementedError("Draft logits are not currently supported in TRTLLMDecoder.")
    
    def _get_embedding_bias(self, logits_type: torch.dtype, tensor: torch.Tensor) -> torch.Tensor:
        """Get embedding bias tensor with correct data type.

        Args:
            logits_type: Target data type for logits
            tensor: Input embedding bias tensor

        Returns:
            Processed embedding bias tensor
        """
        # Return tensor if types match
        if tensor.dtype == logits_type:
            return tensor

        # Handle FP32 to FP16 conversion
        if tensor.dtype == torch.float32 and logits_type == torch.float16:
            return tensor.to(logits_type, device='cuda')

        raise RuntimeError("Embedding bias data type must be same as model logits type.")

File: cpp/tensorrt_llm/batch_manager/generateRequestOptions.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "tensorrt_llm/batch_manager/generateRequestOptions.h"
#include "tensorrt_llm/batch_manager/llmRequest.h"
#include "tensorrt_llm/batch_manager/medusaBuffers.h"
#include "tensorrt_llm/batch_manager/runtimeBuffers.h"
#include "tensorrt_llm/batch_manager/utils/logitsThread.h"
#include "tensorrt_llm/common/logger.h"
#include "tensorrt_llm/common/nvtxUtils.h"

#include <NvInferRuntimeBase.h>

using namespace tensorrt_llm::runtime;

namespace te = tensorrt_llm::executor;

namespace tensorrt_llm::batch_manager
{

std::tuple<ITensor::SharedPtr, std::vector<decoder_batch::Request>, std::vector<SamplingConfig>>
GenerateRequestOptions::operator()(tr::ModelConfig const& modelConfig, tr::WorldConfig const& worldConfig,
    te::DecodingConfig const& decodingConfig, RequestVector const& contextRequests, BufferManager const& bufferManager,
    nvinfer1::DataType logitsType, DecoderInputBuffers& inputBuffers, OptionalRef<RuntimeBuffers const> buffers) const
{
    TLLM_LOG_TRACE("%s start", __PRETTY_FUNCTION__);
    NVTX3_SCOPED_RANGE(GenerateRequestOptions);

    SizeType32 batchSize{0};
    unsigned decoderInputSize{0};
    for (auto const& llmReq : contextRequests)
    {
        auto const& reqTokens = llmReq->getTokens(0);
        if (llmReq->isLastContextChunk())
        {
            decoderInputSize += reqTokens.size();
            ++batchSize;
        }
    }
    inputBuffers.inputsIds->resize(decoderInputSize);

    TensorPtr batchSlotsView = runtime::ITensor::slice(inputBuffers.setupBatchSlots, 0, batchSize);
    auto batchSlotsRange = BufferRange<SizeType32>(*batchSlotsView);
    std::vector<decoder_batch::Request> decoderRequests;
    decoderRequests.reserve(batchSize);
    std::vector<SamplingConfig> samplingConfigs;
    samplingConfigs.reserve(batchSize);

    SizeType32 batchIdx{0};
    SizeType32 inputOffset{0};
    for (auto const& llmReq : contextRequests)
    {
        if (!llmReq->isLastContextChunk())
        {
            continue;
        }

        auto const promptLen = llmReq->getPromptLen();
        auto const& reqTokens = llmReq->getTokens(0);
        TLLM_CHECK(reqTokens.size() == static_cast<decltype(reqTokens.size())>(promptLen));
        TensorPtr inputView = ITensor::slice(inputBuffers.inputsIds, inputOffset, promptLen);
        bufferManager.copy(reqTokens.data(), *inputView);

        auto decoderRequest = decoder_batch::Request{inputView, promptLen, llmReq->mMaxNewTokens, llmReq->mEndId};

        llmReq->mSamplingConfig.normalizeLogProbs = mIsNormalizeLogProbs;
        if (modelConfig.getSpeculativeDecodingMode().isDraftTokensExternal())
        {
            if (llmReq->hasDraftTokens())
            {
                auto const& draftTokens = llmReq->getDraftTokens();
                decoderRequest.draftTokens = bufferManager.copyFrom(*draftTokens, MemoryType::kPINNEDPOOL);
                auto const& draftLogits = llmReq->getDraftLogits();
                if (draftLogits.has_value())
                {
                    decoderRequest.draftLogits
                        = retrieveDraftLogits(modelConfig, worldConfig, draftLogits.value(), bufferManager);
                }
                decoderRequest.generatedTokensPerEngineStep = draftTokens->size() + 1;
            }
            else
            {
                decoderRequest.generatedTokensPerEngineStep = 1;
            }
        }
        else if (!modelConfig.getSpeculativeDecodingMode().isNone())
        {
            decoderRequest.generatedTokensPerEngineStep = modelConfig.getMaxDecodingTokens();
        }
        if (modelConfig.getSpeculativeDecodingMode().isMedusa())
        {
            TLLM_CHECK(buffers);
            llmReq->mSamplingConfig.topKMedusaHeads = {buffers->medusaBuffers->mTopKs};
            // FIXME: we must set medusa paths and tree ids not from seq slot, but from llmRequest?
            // When multiple microbatches buffers are used, runtime buffers can not be addressed with seqSlot.
            decoderRequest.medusaPaths = ITensor::slice(buffers->medusaBuffers->medusaPathsDevice, 0, 1);
            decoderRequest.medusaTreeIds = ITensor::slice(buffers->medusaBuffers->medusaTreeIdsDevice, 0, 1);
        }
        else if (modelConfig.getSpeculativeDecodingMode().isLookaheadDecoding())
        {
            decoderRequest.lookaheadRuntimeConfig = llmReq->getLookaheadConfig()
                ? llmReq->getLookaheadConfig()
                : decodingConfig.getLookaheadDecodingConfig();
        }
        else if (modelConfig.getSpeculativeDecodingMode().isExplicitDraftTokens())
        {
            // Only Explicit draft tokens model needs dtype to WAR the lack of bf16 decoder.
            decoderRequest.dtype = modelConfig.getDataType();
        }
        else if (modelConfig.getSpeculativeDecodingMode().isEagle())
        {
            decoderRequest.eagleConfig
                = llmReq->getEagleConfig() ? llmReq->getEagleConfig() : decodingConfig.getEagleConfig();
        }
        if (llmReq->getEmbeddingBias().has_value())
        {
            decoderRequest.embeddingBias = getEmbeddingBias(logitsType, llmReq->getEmbeddingBias().value());
        }
        if (llmReq->getBadWordsList().has_value())
        {
            // Move to GPU and remove leading bs1 dimension since this is what decoderRequest expects
            decoderRequest.badWordsList = bufferManager.copyFrom(*llmReq->getBadWordsList().value(), MemoryType::kGPU);
            decoderRequest.badWordsList->squeeze(0);
        }
        if (llmReq->getStopWordsList().has_value())
        {
            decoderRequest.stopWordsList
                = bufferManager.copyFrom(*llmReq->getStopWordsList().value(), MemoryType::kGPU);
            decoderRequest.stopWordsList->squeeze(0);
        }
        batchSlotsRange[batchIdx] = llmReq->mSeqSlot.value();
        decoderRequests.push_back(decoderRequest);
        samplingConfigs.push_back(llmReq->mSamplingConfig);

        inputOffset += promptLen;
        ++batchIdx;
    }

    TLLM_LOG_TRACE("%s stop", __PRETTY_FUNCTION__);
    return {std::move(batchSlotsView), std::move(decoderRequests), std::move(samplingConfigs)};
}

std::shared_ptr<runtime::ITensor> GenerateRequestOptions::retrieveDraftLogits(tr::ModelConfig const& modelConfig,
    tr::WorldConfig const& worldConfig, std::shared_ptr<runtime::ITensor> const& tensor,
    BufferManager const& bufferManager) const
{
    TLLM_LOG_TRACE("%s start", __PRETTY_FUNCTION__);

    if (!mSpeculativeDecodingFastLogits)
    {
        TLLM_LOG_TRACE("%s stop", __PRETTY_FUNCTION__);
        return bufferManager.copyFrom(*tensor, MemoryType::kPINNEDPOOL);
    }

    if (mIsLeaderInOrchMode)
    {
        te::SpeculativeDecodingFastLogitsInfo fastLogitsInfo;
        std::memcpy(&fastLogitsInfo, tensor->data(), sizeof(fastLogitsInfo));
        auto logits = utils::targetModelReceiveLogits(fastLogitsInfo, modelConfig).value();

        // Broadcast to other ranks if needed
        if (worldConfig.isTensorParallel())
        {
            auto const& commSession = COMM_SESSION;
            auto shape = logits->getShape();
            commSession.bcastValue(shape.d[0], 0);
            commSession.bcastValue(shape.d[1], 0);
            commSession.bcast(logits->data(), logits->getSizeInBytes(), mpi::MpiType::kUINT8, 0);
        }
        TLLM_LOG_TRACE("%s stop", __PRETTY_FUNCTION__);
        return logits;
    }

    // Get logits from leader rank
    auto const& commSession = COMM_SESSION;
    int64_t dims[2];
    commSession.bcastValue(dims[0], 0);
    commSession.bcastValue(dims[1], 0);
    auto const logitsDtype = modelConfig.getLogitsDtype();
    auto logits = tensorrt_llm::runtime::BufferManager::pinnedPool(ITensor::makeShape({dims[0], dims[1]}), logitsDtype);
    commSession.bcast(logits->data(), logits->getSizeInBytes(), mpi::MpiType::kUINT8, 0);

    TLLM_LOG_TRACE("%s stop", __PRETTY_FUNCTION__);
    return logits;
};

GenerateRequestOptions::TensorPtr GenerateRequestOptions::getEmbeddingBias(
    nvinfer1::DataType logitsType, TensorPtr const& tensor) const
{
    // Check that embedding bias type is same as logits type. If so, we can return the tensor right away
    if (tensor->getDataType() == logitsType)
    {
        return tensor;
    }

    // Support FP32 input for FP16 embedding bias (in the case of FP8 models)
    if (tensor->getDataType() == nvinfer1::DataType::kFLOAT && logitsType == nvinfer1::DataType::kHALF)
    {
        // Do a deep copy of the tensor to the expected type
        TLLM_LOG_WARNING(
            "Embedding bias data type must be same as model logits type, will copy the tensor from float to half");

        TLLM_CHECK_WITH_INFO(
            tensor->getMemoryType() != MemoryType::kGPU, "Embedding bias tensor needs to be in CPU memory for casting");

        auto const shape = tensor->getShape();
        TLLM_CHECK(shape.nbDims == 2); // [1, vocabSizePadded]
        TLLM_CHECK(shape.d[0] == 1);
        auto newTensor = tensorrt_llm::runtime::BufferManager::pinnedPool(shape, logitsType);

        auto const tensorRange = BufferRange<float>(*tensor);
        auto newTensorRange = BufferRange<half>(*newTensor);

        std::transform(tensorRange.begin(), tensorRange.end(), newTensorRange.begin(),
            [](float value) -> half { return static_cast<half>(value); });

        return newTensor;
    }

    TLLM_THROW("Embedding bias data type must be same as model logits type.");
}

} // namespace tensorrt_llm::batch_manager
